src/shared/vscode/commands2.ts

```diff
@@ -367,42 +368,143 @@ interface CommandInfo<T extends Callback> {
     readonly telemetryThrottleMs?: number | false
 }
 
-// This debouncing is a temporary safe-guard to mitigate the risk of instrumenting all
-// commands with telemetry at once. The alternative is to do a gradual rollout although
-// it's not much safer in terms of weeding out problematic commands.
-const telemetryInfo = new Map<string, { startTime: number; debounceCount: number }>()
-
-function getInstrumenter(id: string, threshold: number, telemetryName?: MetricName) {
-    const currentTime = Date.now()
-    const info = telemetryInfo.get(id)
+function getInstrumenter(id: { id: string; args: any[] }, threshold: number, telemetryName?: MetricName) {
+    const currentTime = globals.clock.Date.now()
+    const info = TelemetryDebounceInfo.instance.get(id)
 
+    // to reduce # of events actually emitted, we throttle the same* event for brief period of time
+    // and instead increment a counter for executions while throttled.
     if (!telemetryName && info?.startTime !== undefined && currentTime - info.startTime < threshold) {
         info.debounceCount += 1
-        telemetryInfo.set(id, info)
-        getLogger().debug(`commands: skipped telemetry for "${id}"`)
+        TelemetryDebounceInfo.instance.set(id, info)
+        getLogger().debug(`commands: skipped telemetry for "${id.id}"`)
 
         return undefined
     }
 
     // Throttling occurs regardless of whether or not the instrumenter is invoked
     const span = telemetryName ? telemetry[telemetryName] : telemetry.vscode_executeCommand
     const debounceCount = info?.debounceCount !== 0 ? info?.debounceCount : undefined
-    telemetryInfo.set(id, { startTime: currentTime, debounceCount: 0 })
+    TelemetryDebounceInfo.instance.set(id, { startTime: currentTime, debounceCount: 0 })
 
     return <T extends Callback>(fn: T, ...args: Parameters<T>) =>
         span.run(span => {
-            ;(span as Metric<VscodeExecuteCommand>).record({ command: id, debounceCount })
+            ;(span as Metric<VscodeExecuteCommand>).record({
+                command: id.id,
+                debounceCount,
+                source: BaseCommandSource.findSource(args),
+            })
 
             return fn(...args)
         })
 }
 
+/**
+ * Adding an instance of this to the execution args of your {@link RegisteredCommand}
+ * will set the "source" attribute in the automatically emitted `vscode_executeCommand` metric.
+ */
+export class BaseCommandSource {
```

Comment:
This seems a bit roundabout. We already have a `[RegisteredCommand](https://github.com/aws/aws-toolkit-vscode/blob/df40fccff2f3d55dad8bf9c581f669c7e7edeacf/src/shared/vscode/commands2.ts#L49-L59)` (at least) which is a place where we could define what the "key" of a command is. For expediency, that could be something like:

```diff
diff --git a/src/shared/vscode/commands2.ts b/src/shared/vscode/commands2.ts
index dbb4596cf53d..b8ae76372b08 100644
--- a/src/shared/vscode/commands2.ts
+++ b/src/shared/vscode/commands2.ts
@@ -48,6 +48,7 @@ export interface Command<T extends Callback = Callback> {
  */
 interface RegisteredCommand<T extends Callback = Callback> extends Command<T> {
     dispose(): void
+    setKey(...): void
 }
 
 /**
```

Would this work?

Reply:
Are you implying we now need to refactor the args of `execute()` to separately allow an object with `argName->value` mappings? This would be a lot of work. Otherwise, I may be lacking in that typescript knowledge